ext/reflection/php_reflection.c

```diff
@@ -5127,6 +5127,62 @@ ZEND_METHOD(ReflectionClass, newInstanceArgs)
 }
 /* }}} */
 
+/* {{{ Returns an instance of this class whose properties are filled with the given data before the constructor is called */
+ZEND_METHOD(ReflectionClass, newInstanceFromData)
+{
+	reflection_object *intern;
+	zend_class_entry *ce;
+	int argc = 0;
+	HashTable *data, *args = NULL;
+	zend_function *constructor;
+	zend_string *key;
+	zval *val;
+
+	GET_REFLECTION_OBJECT_PTR(ce);
+
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_ARRAY_HT(data)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_ARRAY_HT(args)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (args) {
+		argc = zend_hash_num_elements(args);
+	}
+
+	if (UNEXPECTED(object_init_ex(return_value, ce) != SUCCESS)) {
+		return;
+	}
+
+	ZEND_HASH_FOREACH_STR_KEY_VAL(data, key, val) {
+		zend_update_property_ex(ce, Z_OBJ_P(return_value), key, val);
+	} ZEND_HASH_FOREACH_END();
+
+	const zend_class_entry *old_scope = EG(fake_scope);
+	EG(fake_scope) = ce;
+	constructor = Z_OBJ_HT_P(return_value)->get_constructor(Z_OBJ_P(return_value));
```

Comment:
This may throw for internal classes.

Reply:
It's the same code used by `newInstance` and `newInstanceArgs` which I'm thinking of refactoring into a shared function to cut the duplication, so would make sense to handle instantiating internal classes there but could be considered a breaking change?